libraries/AP_IOMCU/AP_IOMCU.cpp

```diff
@@ -371,6 +393,66 @@ void AP_IOMCU::read_rc_input()
     }
 }
 
+/*
+  read dshot erpm
+ */
+void AP_IOMCU::read_erpm()
+{
+    uint16_t *r = (uint16_t *)&dshot_erpm;
+    if (!read_registers(PAGE_RAW_DSHOT_ERPM, 0, sizeof(dshot_erpm)/2, r)) {
+        return;
+    }
+    uint8_t motor_poles = 14;
+    AP_BLHeli* blh = AP_BLHeli::get_singleton();
+    if (blh) {
+        motor_poles = blh->get_motor_poles();
+    }
+    for (uint8_t i = 0; i < IOMCU_MAX_CHANNELS/4; i++) {
+        for (uint8_t j = 0; j < 4; j++) {
+            const uint8_t esc_id = (i * 4 + j);
+            if (dshot_erpm.update_mask & 1U<<esc_id) {
+                update_rpm(esc_id, dshot_erpm.erpm[esc_id] * 200U / motor_poles, dshot_telem[i].error_rate[j] / 100.0);
+            }
+        }
+    }
+}
+
+/*
+  read dshot telemetry
+ */
+void AP_IOMCU::read_telem()
+{
+    struct page_dshot_telem* telem = &dshot_telem[esc_group];
+    uint16_t *r = (uint16_t *)telem;
+    iopage page = PAGE_RAW_DSHOT_TELEM_1_4;
+    switch (esc_group) {
+    case 1:
+        page = PAGE_RAW_DSHOT_TELEM_5_8;
+        break;
+    case 2:
+        page = PAGE_RAW_DSHOT_TELEM_9_12;
+        break;
+    case 3:
+        page = PAGE_RAW_DSHOT_TELEM_13_16;
+        break;
+    default:
+        break;
+    }
+
+    if (!read_registers(page, 0, sizeof(page_dshot_telem)/2, r)) {
+        return;
+    }
+    for (uint i = 0; i<4; i++) {
+        TelemetryData t {
+            .temperature_cdeg = int16_t(telem->temperature_cdeg[i]),
+            .voltage = float(telem->voltage_cvolts[i]) * 0.01,
+            .current = float(telem->current_camps[i]) * 0.01
+        };
+        update_telem_data(esc_group * 4 + i, t, telem->types[i]);
+    }
+    esc_group = (esc_group + 1) % 4;
```

Comment:
do we actually support telem from both esc groups? are we wasting half the bandwidth here by requesting telem from a group that can never give it?

Reply:
I have added defines for the telem channels so that we only use what is defined